src/main/java/com/ericsson/eiffel/remrem/semantics/validator/EiffelValidator.java

```diff
@@ -72,41 +88,133 @@ public EiffelValidator(String schemaResourceName, String eventType, List<String>
             throw new IllegalArgumentException(message, e);
         }
     }
-
-    public void validate(JsonObject jsonObjectInput) throws EiffelValidationException {
+    public JsonObject validate(JsonObject jsonObjectInput) throws EiffelValidationException {
+        return validate(jsonObjectInput, false);
+    }
+    
+    public JsonObject validate(JsonObject jsonObjectInput, Boolean lenientValidation) throws EiffelValidationException {
+        JsonArray remremGenerateFailures = new JsonArray();
+        log.debug("VALIDATING Schema used: {}", schemaResourceName);
         try {
-            ProcessingReport report = validationSchema.validate(JsonLoader.fromString(jsonObjectInput.toString()));
-            if (!report.isSuccess()) {
-                log.warn(jsonObjectInput.toString());
-                throw new EiffelValidationException(getErrorsList(report));
+            final ProcessingReport report = validationSchema.validate(JsonLoader.fromString(jsonObjectInput.toString()),
+                    true);
+            if (!report.isSuccess() && lenientValidation) {
+                log.info("Trying to revalidate the Eiffel message with only mandatory Eiffel message fields.");
+                String revalidatedJson = removeErrorProperties(report, jsonObjectInput, remremGenerateFailures);
+                ProcessingReport report2 = validationSchema.validate(JsonLoader.fromString(revalidatedJson), true);
+                handleErrorReport(jsonObjectInput, report2);
+                log.debug("VALIDATED. Schema used: {}", schemaResourceName);
+                return addRemremGenerateFailuresToCustomData(new JsonParser().parse(revalidatedJson).getAsJsonObject(), remremGenerateFailures);
+            } else {
+                handleErrorReport(jsonObjectInput, report);
             }
             log.debug("VALIDATED. Schema used: {}", schemaResourceName);
+            return jsonObjectInput;
         } catch (Exception e) {
-            String message = "Cannot validate given JSON string";
+            final String message = "Cannot validate given JSON string";
             log.debug(message, e);
             throw new EiffelValidationException(message, e);
         }
     }
+    private void handleErrorReport(JsonObject jsonObjectInput, ProcessingReport report) throws EiffelValidationException {
+        if (!report.isSuccess()) {
+            log.error("Eiffel message validation failed");
+            log.error(jsonObjectInput.toString());
+            throw new EiffelValidationException(getErrorsList(report));
```

Comment:
The strategy for error logging and exception handling seems haphazard. Logging _and_ throwing an exception from the same place usually isn't a great idea (you're likely to log duplicate information, plus it's makes the code verbose). Here you're logging the input message at the error level with no information about the error itself, throwing an exception with the errors, which is caught on line 113 and logged at the debug level, then re-raised as a new exception that wraps the original exception. Perhaps the exception from line 116 is also logged somewhere?

Wouldn't it be better to raise an exception that contains all relevant information and log that in _one_ place?

But I realize most of this is old code that shouldn't be modified in this PR.

Also, when logging, please don't call the logger multiple times for what's effectively a single message (like the case with lines 121-122).

Reply:
Yes, I agree loggers are not required here, we are raising the exception with validation errors.